be/src/exec/pipeline/scan/connector_scan_operator.cpp

```diff
@@ -216,44 +458,64 @@ Status ConnectorChunkSource::_open_data_source(RuntimeState* state) {
 }
 
 Status ConnectorChunkSource::_read_chunk(RuntimeState* state, ChunkPtr* chunk) {
-    RETURN_IF_ERROR(_open_data_source(state));
+    ConnectorScanOperatorAdaptiveProcessor& P = *(_op->_adaptive_processor);
 
-    if (state->is_cancelled()) {
-        return Status::Cancelled("canceled state");
-    }
+    DeferOp op([&]() { P.last_chunk_souce_finish_timestamp = GetCurrentTimeMicros(); });
 
-    // Improve for select * from table limit x, x is small
-    if (_reach_eof()) {
-        return Status::EndOfFile("limit reach");
-    }
+    int64_t total_time_ns = 0;
+    int64_t delta_io_time_ns = 0;
+    int64_t delta_scan_bytes = 0;
+    {
+        SCOPED_RAW_TIMER(&total_time_ns);
+        int64_t prev_io_time_ns = get_io_time_spent();
+        int64_t prev_scan_bytes = get_scan_bytes();
 
-    while (_status.ok()) {
-        ChunkPtr tmp;
-        _status = _data_source->get_next(state, &tmp);
-        if (_status.ok()) {
-            if (tmp->num_rows() == 0) continue;
-            _ck_acc.push(tmp);
-            if (_ck_acc.has_output()) break;
-        } else if (!_status.is_end_of_file()) {
-            if (_status.is_time_out()) {
-                Status t = _status;
-                _status = Status::OK();
-                return t;
+        RETURN_IF_ERROR(_open_data_source(state));
+        if (state->is_cancelled()) {
+            return Status::Cancelled("canceled state");
+        }
+
+        // Improve for select * from table limit x, x is small
+        if (_reach_eof()) {
+            return Status::EndOfFile("limit reach");
+        }
+
+        while (_status.ok()) {
+            ChunkPtr tmp;
+            _status = _data_source->get_next(state, &tmp);
+            if (_status.ok()) {
+                if (tmp->num_rows() == 0) continue;
+                _ck_acc.push(tmp);
+                if (_ck_acc.has_output()) break;
+            } else if (!_status.is_end_of_file()) {
+                if (_status.is_time_out()) {
+                    Status t = _status;
+                    _status = Status::OK();
+                    return t;
+                } else {
+                    return _status;
+                }
             } else {
-                return _status;
+                _ck_acc.finalize();
+                DCHECK(_status.is_end_of_file());
             }
-        } else {
-            _ck_acc.finalize();
-            DCHECK(_status.is_end_of_file());
         }
+
+        DCHECK(_status.ok() || _status.is_end_of_file());
+        _scan_rows_num = _data_source->raw_rows_read();
+        _scan_bytes = _data_source->num_bytes_read();
+        _cpu_time_spent_ns = _data_source->cpu_time_spent();
+        _io_time_spent_ns = _data_source->io_time_spent();
```

Comment:
why not record them in a thread local variable and collect them when you actually need them?

Reply:
because data source/chunk source can be run on any thread. 

even different query's chunk sources can be run on any thread.

Reply:
but seems we only need aggregated results, so no matter which thread they are running, we can aggregate them from different threads